source/common/access_log/access_log_formatter.cc

```diff
@@ -61,12 +61,15 @@ FormatterPtr AccessLogFormatUtils::defaultAccessLogFormatter() {
   return FormatterPtr{new FormatterImpl(DEFAULT_FORMAT)};
 }
 
-const std::string&
-AccessLogFormatUtils::protocolToString(const absl::optional<Http::Protocol>& protocol) {
-  if (protocol) {
+const std::string AccessLogFormatUtils::protocolToString(absl::Span<const std::string> protocols) {
+  auto protocol = Http::Utility::getProtocol(protocols);
+  if (protocol.has_value()) {
```

Comment:
Why do we need to do the string to enum to string dance here? Can't we just return the protocol string directly? Or this is for HTTP back compat? If for back compat can you add comments and also maybe skip  the final  `getProtocolString` call by returning the string reference directly from `getProtocol`?

Reply:
Mostly for backwards compatibility. I can change the interface to return both the enum and the string.

Reply:
I think trying to mix strings and the protocol enum is going to be confusing. It seems like it will be unavoidable to have an enum interface method that can only return a subset of the protocols encoded as strings.

I think the approach needs to either 1) use strings every where (and convert from the protocol enum to strings only once for each protocol) or else 2) switch everything to a mechanism that has a registry-style extension mechanism and pass around those values and convert them to strings at the point where we need a string.

Reply:
Got it. I'll preserve protocol() for HTTP as-is and add two more stubs to deal with the layered protocols. Thank you!

Reply:
+1 this sounds like a good plan. Thank you!